src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackDecoder.cs

```diff
@@ -260,7 +260,16 @@ public void Decode(ReadOnlySpan<byte> data, HeaderCallback onHeader, object onHe
 
                         if (_integerDecoder.StartDecode((byte)(b & ~HuffmanMask), StringLengthPrefix))
                         {
-                            OnStringLength(_integerDecoder.Value, nextState: State.HeaderValue);
+                            if (_integerDecoder.Value > 0)
+                            {
+                                OnStringLength(_integerDecoder.Value, nextState: State.HeaderValue);
+                            }
+                            else
+                            {
+                                OnString(nextState: State.Ready);
```

Comment:
Does this actually work? It seems like OnString is depending on _stringLength being set properly (i.e. to 0 for this case) but since we skip the OnStringLength call, it looks like _stringLength won't be set here.

Reply:
Also, it looks like we are basically duplicating the logic from the State.HeaderValue case here, when we reach the end of the string. Can we refactor so there's a common method they both call?

Reply:
And it looks like we're ignoring the _index flag here, whereas we handle it properly in the State.HeaderValue case. Another reason to have a common routine.

Reply:
It is updated now. The old code did work but it did unnecessary copy with old value. 

I did run tools @stephentoub wrote and 12 out of 13 failing sites works now (returns 200)